api/proto/teleport/legacy/types/types.proto

```diff
@@ -4743,3 +4743,106 @@ enum RequireMFAType {
   // requiring double touch on session requests, normal Session MFA is disabled.
   HARDWARE_KEY_TOUCH = 3;
 }
+
+// Plugin describes a single instance of a Teleport Plugin
+message PluginV1 {
+  // Kind is the plugin resource kind.
+  string Kind = 1 [(gogoproto.jsontag) = "kind"];
+  // SubKind is an optional resource subkind.
+  string SubKind = 2 [(gogoproto.jsontag) = "sub_kind,omitempty"];
+  // Version is the resource version.
+  string Version = 3 [(gogoproto.jsontag) = "version"];
+
+  // Metadata is the resource metadata.
+  Metadata Metadata = 4 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "metadata"
+  ];
+
+  PluginSpecV1 Spec = 5 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "spec"
+  ];
+
+  PluginStatusV1 Status = 6 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "status"
+  ];
+
+  // Credentials are "live" credentials to the 3rd party API.
+  // These are considered secrets.
+  PluginCredentialsV1 Credentials = 7 [(gogoproto.jsontag) = "credentials,omitempty"];
+}
+
+message PluginSpecV1 {
+  option (gogoproto.equal) = true;
+
+  // Settings contain provider-specific plugin options.
+  oneof Settings {
+    // Settings for the Slack access plugin
+    PluginSlackAccessSettings SlackAccessPlugin = 1 [(gogoproto.jsontag) = "slack_access,omitempty"];
+  }
+}
+
+message PluginSlackAccessSettings {
+  option (gogoproto.equal) = true;
+
+  string FallbackChannel = 1 [(gogoproto.jsontag) = "fallback_channel"];
+}
+
+message PluginInitialCredentialsV1 {
+  oneof Credentials {
+    PluginOAuth2AuthorizationCodeCredentials OAuth2AuthorizationCode = 1 [(gogoproto.jsontag) = "oauth2_authorization_code,omitempty"];
+  }
+}
+
+message PluginOAuth2AuthorizationCodeCredentials {
+  string AuthorizationCode = 1 [(gogoproto.jsontag) = "authorization_code"];
+  string RedirectURI = 2 [(gogoproto.jsontag) = "redirect_uri"];
+}
+
+// PluginStatus is the user-facing status for the plugin instance.
+message PluginStatusV1 {
+  PluginStatusCode Code = 1 [(gogoproto.jsontag) = "code"];
+}
```

Comment:
Having a code and a message (user-facing string) could be useful.

Reply:
I was thinking about a message field, however for now the error conditions seem to be simple, we could probably map each code to a human-readable string 1-to-1 (whether on the backend, or on the frontend)?